File: backend/src/appattest.ts
// Appattest endpoint
// #[post("/appattest", format = "application/json", data = "<attestation_data>")]
// async fn appattest(attestation_data: Json<AttestationData<'_>>) -> () {
//     const APP_ID: &str = "proof-pix";

//     // Add challenge to used challenges
//     let added = add_challenge(&con, attestation_data.challenge)
//         .await
//         .expect("Failed to add challenge");

//     if added {
//         println!("Challenge added.");
//     } else {
//         println!("Challenge already exists.");
//         return;
//     }

//     let verified = app_attest::validate_raw_attestation(
//         attestation_data.attestation_string,
//         attestation_data.challenge,
//         attestation_data.raw_key_id,
//         APP_ID,
//         false, // production
//         false, // leaf_cert_only
//     );

//     // If verified
//     if verified {
//         println!("Verified attestation");
//         // can do something
//     }
// }

File: backend/src/appattest.rs
use dotenv::dotenv;
use rocket::form::FromForm;
use std::env;

extern crate rocket;

use crate::db::add_challenge;

// #[derive(Deserialize)]
// #[serde(crate = "rocket::serde")]
#[derive(FromForm, Debug)]
pub struct AttestationData {
    pub attestation_string: String,
    pub raw_key_id: String,
    pub challenge: String,
}

// validate_attestation
pub async fn validate_attestation(attestation_data: AttestationData) -> &'static str {
    dotenv().ok();
    let app_id = env::var("APP_ID").expect("APP_ID must be set");
    let dev: String = env::var("DEV").expect("DEV must be set");
    let prod = if dev == "true" { false } else { true };

    let added = add_challenge(
        attestation_data.challenge.clone(),
        attestation_data.attestation_string.clone(),
    )
    .await;

    if added {
        println!("Challenge added.");
    } else {
        println!("Challenge already exists. Invalid attestation.");
        return "challenge already exists";
    }

    // Verify the attestation
    println!("Verifying attestation... {:?}", attestation_data);
    let verified = app_attest::validate_raw_attestation(
        &attestation_data.attestation_string,
        &attestation_data.raw_key_id,
        &attestation_data.challenge,
        &app_id,
        prod,  // production
        false, // leaf_cert_only
    );

    // If verified
    if verified {
        return "Verified attestation";
    } else {
        return "Invalid attestation";
    }
}
